src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs

```diff
@@ -3504,6 +3505,56 @@ private BoundExpression BindNullCoalescingOperator(BinaryExpressionSyntax node,
             return GenerateNullCoalescingBadBinaryOpsError(node, leftOperand, rightOperand, Conversion.NoConversion, diagnostics);
         }
 
+
+        private BoundExpression BindNullCoalescingAssignmentOperator(AssignmentExpressionSyntax node, DiagnosticBag diagnostics)
+        {
+            BoundExpression leftOperand = BindValue(node.Left, diagnostics, BindValueKind.CompoundAssignment);
+            BoundExpression rightOperand = BindValue(node.Right, diagnostics, BindValueKind.RValue);
+
+            // If either operand is bad, bail out preventing more cascading errors
+            if (leftOperand.HasAnyErrors || rightOperand.HasAnyErrors)
+            {
+                return new BoundNullCoalesingAssignmentOperator(node, leftOperand, rightOperand, CreateErrorType(), hasErrors: true);
+            }
+
+            // Unlike a standard null coalescing expression, the resulting type of the expression a ??= b
+            // must be A (where A is the type of a), as it is where the result of the assignment will end up. Therefore,
+            // we must ensure that B (where B is the type B) is implicitly convertible to A or A0 (where A0 is the
+            // underlying non-nullable type of A).
+            TypeSymbol leftType = leftOperand.Type;
+            Debug.Assert(leftType != null);
+            TypeSymbol optRightType = rightOperand.Type;
+            bool rightTypeIsNonNull = !(optRightType is null);
+            bool isLeftNullable = leftType.IsNullableType();
+
+            // If A exists and is not a nullable type or reference type, a compile-time error occurs
+            if (!leftType.IsReferenceType && !isLeftNullable)
+            {
+                return generateNullCoalescingAssignmentBadBinaryOpsError();
+            }
+
+            // If A exists and an implicit conversion exists from B to A, we store that conversion. At runtime,
+            // a is first evaluated. If a is not null, b is not evaluated. If a is null, b is evaluated and converted
+            // to type A, and is stored in a.
+            HashSet<DiagnosticInfo> useSiteDiagnostics = null;
+            var rightConversion = Conversions.ClassifyImplicitConversionFromExpression(rightOperand, leftType, ref useSiteDiagnostics);
+            diagnostics.Add(node, useSiteDiagnostics);
+            if (rightConversion.Exists)
+            {
+                var convertedRightOperand = CreateConversion(rightOperand, rightConversion, leftType, diagnostics);
+                return new BoundNullCoalesingAssignmentOperator(node, leftOperand, convertedRightOperand, leftType);
+            }
+
+            // a and b are incompatible and a compile-time error occurs
+            return generateNullCoalescingAssignmentBadBinaryOpsError();
+
+            BoundExpression generateNullCoalescingAssignmentBadBinaryOpsError()
```

Comment:
>generateNullCoalescingAssignmentBadBinaryOpsError [](start = 28, length = 49)

Consider creating an instance method rather than a local function, for consistency with the `BoundNullCoalescingOperator` case.

Reply:
It seemed to me the only reason that BoundNullCoalescingOperator's version isn't a local method is that such didn't exist when that was created. I could refactor that to be a local method if you'd prefer.

---
In reply to: [199842701](https://github.com/dotnet/roslyn/pull/28253#discussion_r199842701) [](ancestors = 199842701)

Reply:
We should avoid unnecessary style changes of existing code, particularly in branches where there may be many iterations and merges before the branch is submitted. And in cases like this where similar code is added, we should match the existing style.

---
In reply to: [199902410](https://github.com/dotnet/roslyn/pull/28253#discussion_r199902410) [](ancestors = 199902410,199842701)